src/_zkapauthorizer/model.py

```diff
@@ -439,6 +442,64 @@ def insert_unblinded_tokens_for_voucher(
                 for token in unblinded_tokens
             ),
         )
+        self._delete_corresponding_tokens(cursor, voucher, unblinded_tokens)
+
+    def _delete_corresponding_tokens(self, cursor, voucher : bytes, unblinded_tokens : List["UnblindedToken"]) -> None:
+        """
+        Delete rows from the [tokens] table corresponding to the given unblinded
+        tokens.
+        """
+        # The only way to match tokens with unblinded tokens is to compare the
+        # preimages they each contain.  Unfortunately this means we have to
+        # load all of the tokens from the database.  Hopefully this will never
+        # be a truly huge number because we clean up the table as we make
+        # progress on redemption.
```

Comment:
While this is true, looking at the code for `BatchDLEQProof..invalid_or_unblind`, it looks like the code always returns the list of unblinded tokens in the same order as the tokens that were passed in[1]. Thus, it would be possible for the caller to pass in the corresponding list of tokens, rather than requiring a bunch of logic here to determine the list of tokens to delete.

Specifically, if `random_tokens` were passed to `RedemptionResult` [here](https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ccc0d28fb4f54c10722ee311945073296e557fec/src/_zkapauthorizer/controller.py#L575-L578), they could be passed to `insert_unblinded_tokens_for_voucher` [here](https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ccc0d28fb4f54c10722ee311945073296e557fec/src/_zkapauthorizer/controller.py#L918-L924), which could then be passed into this function.

[1] I think it is reasonable to depend on this, since that function also depends on the list of blinded and signed tokens to be in the same order. And, in any case, it doesn't actually appear that we would be depending on the order.

---

Alternatively, it looks like the tokens to delete are all the tokens in the database associated with the given vouncher and counter, so, if the counter was passed into `insert_unblinded_tokens_for_voucher` [here](https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ccc0d28fb4f54c10722ee311945073296e557fec/src/_zkapauthorizer/controller.py#L918-L924), we could just delete tokens based on that info. Looking at 
['PaymentController._get_random_tokens_for_voucher`](https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ccc0d28fb4f54c10722ee311945073296e557fec/src/_zkapauthorizer/controller.py#L767-L801)
and [`VoucherStore.add`](https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ccc0d28fb4f54c10722ee311945073296e557fec/src/_zkapauthorizer/model.py#L264-L339) the tokens that will be deleted by the code in this PR will always match the tokens with the given voucher and counter.

Reply:
Yea good call.  This was actually plan A (94f8ca027a4a12e1c9543396f90bd274cf7b8c09) but it looks like I picked the wrong column name and then (mistakenly) convinced myself the current schema wouldn't support that approach.

I've switched the branch back to that idea w/ a correct implementation - with one variation on your suggestion.  I just pulled the counter out of the database instead of requiring it as a parameter to `insert_unblinded_tokens_for_voucher`.  The database state has to be consistent with the API calls being made (so the value used should be the same either way) and this way I don't have to update all the callers of `insert_unblinded_tokens_for_voucher`.